titus-common/src/main/java/com/netflix/titus/common/util/loadshedding/tokenbucket/TokenBucketAdmissionController.java

```diff
@@ -0,0 +1,146 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.common.util.loadshedding.tokenbucket;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.cache.Cache;
+import com.netflix.titus.common.util.cache.Caches;
+import com.netflix.titus.common.util.limiter.Limiters;
+import com.netflix.titus.common.util.limiter.tokenbucket.TokenBucket;
+import com.netflix.titus.common.util.loadshedding.AdmissionController;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerRequest;
+import com.netflix.titus.common.util.loadshedding.AdmissionControllerResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Admission controller with multiple token buckets. A token bucket is selected by evaluating matching criteria in
+ * declaration order until first match is found. The selected bucket is tried, and either success or failure is
+ * returned to the caller.
+ */
+public class TokenBucketAdmissionController implements AdmissionController {
+
+    private static final Logger logger = LoggerFactory.getLogger(TokenBucketAdmissionController.class);
+
+    private static final AdmissionControllerResponse DEFAULT_OK = AdmissionControllerResponse.newBuilder()
+            .withAllowed(true)
+            .withReasonMessage("Rate limits not configured")
+            .build();
+
+    private static final int MAX_CACHE_SIZE = 10_000;
+    private static final Duration CACHE_ITEM_TIMEOUT = Duration.ofSeconds(60);
+
+    private final List<TokenBucketConfiguration> tokenBucketConfigurations;
+    private final Cache<String, TokenBucketInstance> bucketsById;
+
+    public TokenBucketAdmissionController(List<TokenBucketConfiguration> tokenBucketConfigurations,
+                                          TitusRuntime titusRuntime) {
+        this.tokenBucketConfigurations = tokenBucketConfigurations;
+
+        this.bucketsById = Caches.instrumentedCacheWithMaxSize(
+                MAX_CACHE_SIZE,
+                CACHE_ITEM_TIMEOUT,
+                "titus.tokenBucketAdmissionController.cache",
+                titusRuntime.getRegistry()
+        );
+    }
+
+    @Override
+    public AdmissionControllerResponse apply(AdmissionControllerRequest request) {
+        return findTokenBucket(request).map(this::consume).orElse(DEFAULT_OK);
+    }
+
+    private Optional<TokenBucketInstance> findTokenBucket(AdmissionControllerRequest request) {
+        return tokenBucketConfigurations.stream()
```

Comment:
I don't particularly love the `O(N)` here on every single request, I think it will be fine as the number of `tokenBucketConfigurations` will be very small, but we should probably monitor latency here to make sure we keep an eye on it

Reply:
My assumption is that we should have <10 configurations total. If we need more and it gets more complex, we should optimize this code.

Reply:
I would have rather seen a generated key + lookup to make this more efficient but it isn't needed immediately.

Reply:
Giving up under the peer pressure. I am implementing the caching now.